File: tests/api_test.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"sync"
	"testing"
	"time"
)

const (
	APIBase    = "http://localhost:3000/api/fitness/v1"
	TestUserID = "test-user-123"
	AuthToken  = "mock-jwt-token"
)

type DataSource struct {
	DataStreamName string `json:"dataStreamName"`
	Type           string `json:"type"`
	Application    struct {
		PackageName string `json:"packageName"`
		Version     string `json:"version"`
	} `json:"application"`
	DataType struct {
		Name  string `json:"name"`
		Field []struct {
			Name   string `json:"name"`
			Format string `json:"format"`
		} `json:"field"`
	} `json:"dataType"`
}

type Session struct {
	Name             string `json:"name"`
	Description      string `json:"description"`
	StartTimeMillis  int64  `json:"startTimeMillis"`
	EndTimeMillis    int64  `json:"endTimeMillis"`
	ActivityType     int    `json:"activityType"`
	Application      struct {
		PackageName string `json:"packageName"`
	} `json:"application"`
}

func makeRequest(method, url string, body interface{}) (*http.Response, error) {
	var reqBody io.Reader
	if body != nil {
		jsonBody, _ := json.Marshal(body)
		reqBody = bytes.NewBuffer(jsonBody)
	}

	req, err := http.NewRequest(method, url, reqBody)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Authorization", "Bearer "+AuthToken)
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: 10 * time.Second}
	return client.Do(req)
}

func TestHealthyDuckAPI(t *testing.T) {
	t.Run("DataSourceCreation", func(t *testing.T) {
		dataSource := DataSource{
			DataStreamName: "com.ultimatequack.heart_rate",
			Type:          "raw",
		}
		dataSource.Application.PackageName = "com.ultimatequack.healthyduck"
		dataSource.Application.Version = "1.0.0"
		dataSource.DataType.Name = "com.ultimatequack.heart_rate"
		dataSource.DataType.Field = []struct {
			Name   string `json:"name"`
			Format string `json:"format"`
		}{
			{Name: "bpm", Format: "floatPoint"},
		}

		url := fmt.Sprintf("%s/users/%s/dataSources", APIBase, TestUserID)
		resp, err := makeRequest("POST", url, dataSource)
		if err != nil {
			t.Fatalf("Failed to create data source: %v", err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != 201 {
			t.Errorf("Expected status 201, got %d", resp.StatusCode)
		}
	})

	t.Run("SessionManagement", func(t *testing.T) {
		session := Session{
			Name:            "HIIT Workout",
			Description:     "High-intensity interval training",
			StartTimeMillis: time.Now().Add(-time.Hour).UnixMilli(),
			EndTimeMillis:   time.Now().UnixMilli(),
			ActivityType:    79, // HIIT
		}
		session.Application.PackageName = "com.ultimatequack.healthyduck"

		url := fmt.Sprintf("%s/users/%s/sessions", APIBase, TestUserID)
		resp, err := makeRequest("POST", url, session)
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != 201 {
			t.Errorf("Expected status 201, got %d", resp.StatusCode)
		}
	})
}

func BenchmarkAPIPerformance(b *testing.B) {
	b.Run("DataSourceRetrieval", func(b *testing.B) {
		url := fmt.Sprintf("%s/users/%s/dataSources", APIBase, TestUserID)
		b.ResetTimer()
		
		for i := 0; i < b.N; i++ {
			resp, err := makeRequest("GET", url, nil)
			if err != nil {
				b.Fatalf("Request failed: %v", err)
			}
			resp.Body.Close()
		}
	})

	b.Run("ConcurrentRequests", func(b *testing.B) {
		url := fmt.Sprintf("%s/users/%s/dataSources", APIBase, TestUserID)
		b.ResetTimer()

		b.RunParallel(func(pb *testing.PB) {
			for pb.Next() {
				resp, err := makeRequest("GET", url, nil)
				if err != nil {
					b.Errorf("Request failed: %v", err)
					continue
				}
				resp.Body.Close()
			}
		})
	})
}

func TestConcurrentDataInsertion(t *testing.T) {
	const numGoroutines = 10
	const requestsPerGoroutine = 5

	var wg sync.WaitGroup
	errors := make(chan error, numGoroutines*requestsPerGoroutine)

	for i := 0; i < numGoroutines; i++ {
		wg.Add(1)
		go func(goroutineID int) {
			defer wg.Done()
			
			for j := 0; j < requestsPerGoroutine; j++ {
				dataSource := DataSource{
					DataStreamName: fmt.Sprintf("com.ultimatequack.test_%d_%d", goroutineID, j),
					Type:          "raw",
				}
				dataSource.Application.PackageName = "com.ultimatequack.healthyduck"
				dataSource.Application.Version = "1.0.0"

				url := fmt.Sprintf("%s/users/%s/dataSources", APIBase, TestUserID)
				resp, err := makeRequest("POST", url, dataSource)
				if err != nil {
					errors <- fmt.Errorf("goroutine %d, request %d failed: %v", goroutineID, j, err)
					continue
				}
				resp.Body.Close()

				if resp.StatusCode != 201 {
					errors <- fmt.Errorf("goroutine %d, request %d: expected status 201, got %d", goroutineID, j, resp.StatusCode)
				}
			}
		}(i)
	}

	wg.Wait()
	close(errors)

	errorCount := 0
	for err := range errors {
		t.Error(err)
		errorCount++
	}

	if errorCount > 0 {
		t.Errorf("Concurrent test failed with %d errors", errorCount)
	}
}

File: tests/api.test.js
const axios = require("axios")

const API_BASE = process.env.API_BASE || "http://localhost:3000"
const TEST_USER_ID = "test-user-123"

// Mock authentication token for testing
const AUTH_TOKEN = "mock-jwt-token"

const apiClient = axios.create({
  baseURL: `${API_BASE}/api/fitness/v1`,
  headers: {
    Authorization: `Bearer ${AUTH_TOKEN}`,
    "Content-Type": "application/json",
  },
})

describe("HealthyDuck Fitness API - JavaScript Integration Tests", () => {
  let testDataSourceId
  let testSessionId

  beforeAll(async () => {
    console.log("Setting up JavaScript API tests...")
  })

  afterAll(async () => {
    console.log("Cleaning up JavaScript API tests...")
  })

  describe("Data Sources API", () => {
    test("should create a new data source", async () => {
      const dataSource = {
        dataStreamName: "com.ultimatequack.step_count",
        type: "derived",
        application: {
          packageName: "com.ultimatequack.healthyduck",
          version: "1.0.0",
        },
        dataType: {
          name: "com.ultimatequack.step_count",
          field: [{ name: "steps", format: "integer" }],
        },
      }

      const response = await apiClient.post(`/users/${TEST_USER_ID}/dataSources`, dataSource)
      expect(response.status).toBe(201)
      expect(response.data.dataStreamId).toBeDefined()
      testDataSourceId = response.data.dataStreamId
    })

    test("should retrieve data sources", async () => {
      const response = await apiClient.get(`/users/${TEST_USER_ID}/dataSources`)
      expect(response.status).toBe(200)
      expect(Array.isArray(response.data.dataSource)).toBe(true)
    })

    test("should handle fitness data insertion", async () => {
      const dataset = {
        dataSourceId: testDataSourceId,
        maxEndTimeNs: Date.now() * 1000000,
        minStartTimeNs: (Date.now() - 3600000) * 1000000,
        point: [
          {
            startTimeNanos: (Date.now() - 1800000) * 1000000,
            endTimeNanos: Date.now() * 1000000,
            dataTypeName: "com.ultimatequack.step_count",
            value: [{ intVal: 5000 }],
          },
        ],
      }

      const response = await apiClient.patch(
        `/users/${TEST_USER_ID}/dataSources/${testDataSourceId}/datasets/${dataset.minStartTimeNs}-${dataset.maxEndTimeNs}`,
        dataset,
      )
      expect(response.status).toBe(200)
    })
  })

  describe("Sessions API", () => {
    test("should create workout session", async () => {
      const session = {
        name: "Morning Run",
        description: "Daily cardio workout",
        startTimeMillis: Date.now() - 3600000,
        endTimeMillis: Date.now(),
        activityType: 8, // Running
        application: {
          packageName: "com.ultimatequack.healthyduck",
        },
      }

      const response = await apiClient.post(`/users/${TEST_USER_ID}/sessions`, session)
      expect(response.status).toBe(201)
      testSessionId = response.data.id
    })

    test("should retrieve user sessions", async () => {
      const response = await apiClient.get(`/users/${TEST_USER_ID}/sessions`)
      expect(response.status).toBe(200)
      expect(Array.isArray(response.data.session)).toBe(true)
    })
  })

  describe("Data Aggregation", () => {
    test("should aggregate daily fitness data", async () => {
      const endTime = Date.now()
      const startTime = endTime - 24 * 60 * 60 * 1000 // 24 hours ago

      const response = await apiClient.get(`/users/${TEST_USER_ID}/dataset/aggregate/daily`, {
        params: {
          startTimeMillis: startTime,
          endTimeMillis: endTime,
          dataTypeName: "com.ultimatequack.step_count",
        },
      })

      expect(response.status).toBe(200)
      expect(response.data.bucket).toBeDefined()
    })
  })

  describe("Error Handling", () => {
    test("should handle unauthorized requests", async () => {
      const unauthorizedClient = axios.create({
        baseURL: `${API_BASE}/api/fitness/v1`,
      })

      try {
        await unauthorizedClient.get(`/users/${TEST_USER_ID}/dataSources`)
      } catch (error) {
        expect(error.response.status).toBe(401)
      }
    })

    test("should handle invalid data source creation", async () => {
      const invalidDataSource = {
        dataStreamName: "", // Invalid empty name
        type: "invalid_type",
      }

      try {
        await apiClient.post(`/users/${TEST_USER_ID}/dataSources`, invalidDataSource)
      } catch (error) {
        expect(error.response.status).toBe(400)
      }
    })
  })
})
